soccer_ipm/soccer_ipm/soccer_ipm.py

```diff
@@ -0,0 +1,323 @@
+        field = self.get_field(msg.header.stamp, 0.0)
+
+        field_boundary = sv3dm.FieldBoundary()
+        field_boundary.header.stamp = msg.header.stamp
+        field_boundary.header.frame_id = self._base_footprint_frame
+        field_boundary.confidence = field_boundary.confidence
+
+        for p in msg.points:
+            image_point = Point2DStamped(
+                header=msg.header,
+                point=p)
+            # Map point from image onto field plane
+            try:
+                relative_foot_point = self.ipm.map_point(
+                    field,
+                    image_point,
+                    output_frame=self._base_footprint_frame)
+
+                field_boundary.points.append(relative_foot_point.point)
+            except NoIntersectionError:
+                self.get_logger().warn(
+                    'Got a field boundary point ({},{}) I could not transform.'.format(
+                        image_point.point.x,
+                        image_point.point.y),
+                    throttle_duration_sec=5)
+
+        self.field_boundary_pub.publish(field_boundary)
+
+    def callback_masks(self, msg: Image, publisher, encoding='8UC1', scale: float = 1.0):
+        """
+        Map a mask from the input image as a pointcloud on the field plane.
+
+        :param msg: Mask msg type
+        :param publisher: Publisher which should be used to publish the mapped point cloud
+        :param encoding: Encoding of the input mask. For the exact format see the cv_bride docs.
+        :param scale: Downsampling which is applied to the mask before the mapping.
+        """
+        # Get field plane
+        field = self.get_field(msg.header.stamp, 0.0)
+        if field is None:
+            return
+
+        # Convert subsampled image
+        image = cv2.resize(
+            self._cv_bridge.imgmsg_to_cv2(msg, encoding),
+            (0, 0), fx=scale, fy=scale, interpolation=cv2.INTER_NEAREST)
+
+        # Get indices for all non 0 pixels (the pixels which should be displayed in the pointcloud)
+        point_idx_tuple = np.where(image != 0)
+
+        # Restructure index tuple to a array
+        point_idx_array = np.empty((point_idx_tuple[0].shape[0], 3))
+        point_idx_array[:, 0] = point_idx_tuple[1] / scale
+        point_idx_array[:, 1] = point_idx_tuple[0] / scale
+
+        # Map points
+        points_on_plane = self.ipm.map_points(
+                    field,
+                    point_idx_array,
+                    msg.header,
+                    output_frame=self._base_footprint_frame)
+
+        # Make a pointcloud2 out of them
+        pc = create_cloud_xyz32(
+            Header(
+                stamp=msg.header.stamp,
+                frame_id=self._base_footprint_frame
+            ),
+            points_on_plane)
+
+        # Publish point cloud
+        publisher.publish(pc)
+
+    def _object_at_bottom_of_image(self, position, thresh):
```

Comment:
This private method is confusing me. What's the point of thresh? (Setting this to 0.8 would consider any pixels in the bottom 10% as out of the image?)
Ideally, I'd like this private method split out into a function in a separate file so we write some.

Reply:
There is a problem with the ipm method. If the footpoint of a given object is not in the image, and you observe the object e.g. as a bounding box in image space, you might falsely assume that the lowest part of the bounding box is the footpoint. A simple but effective way of mitigating this is to assume that all bounding boxes that end at the bottom of the image might continue to a lower footpoint. They are therefore not projectable and dropped. Otherwise, some weird "alias" projections might occur. Because object detectors like YOLO are not 100% accurate, boxes of objects that continue might end in the lower part of the image instead of exactly the lowest pixel. This method is used to threshold this and sort out objects that would be projected to a wrong position (really messes with filters and stuff) otherwise.

Reply:
> Ideally, I'd like this private method split out into a function in a separate file so we write some.

I don't really understand what you are saying. ^^

Reply:
> I don't really understand what you are saying here. ^^

That statement should've been _Ideally, I'd like this private method split out into a function in a separate file so we write some tests._

Reply:
> There is a problem with the ipm method. If the footpoint of a given object is not in the image, and you observe the object e.g. as a bounding box in image space, you might falsely assume that the lowest part of the bounding box is the footpoint. A simple but effective way of mitigating this is to assume that all bounding boxes that end at the bottom of the image might continue to a lower footpoint. They are therefore not projectable and dropped. Otherwise, some weird "alias" projections might occur. Because object detectors like YOLO are not 100% accurate, boxes of objects that continue might end in the lower part of the image instead of exactly the lowest pixel. This method is used to threshold this and sort out objects that would be projected to a wrong position (really messes with filters and stuff) otherwise.

I see, I think this is quite a big issue that should be discussed further outside this PR.

Reply:
Okay